File: src/backend/v2/apps/user-auth/src/app/auth/auth.service.ts
import {
  BadRequestException,
  ForbiddenException,
  Injectable,
  NotFoundException,
  UnauthorizedException,
} from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { User } from '../user/entities/user.entity';
import { Repository } from 'typeorm';
import { JwtService } from '@nestjs/jwt';
import { InjectPinoLogger, PinoLogger } from 'nestjs-pino';
import { Bcrypt } from '@firewall-backend/utils';
import { ResetPasswordDto } from './dto/auth.dto';
import { LoginDto } from './dto/auth.dto';
import { EULA } from '../eula/entities/eula.entity';

@Injectable()
export class AuthService {
  constructor(
    @InjectRepository(User)
    private readonly userRepository: Repository<User>,
    @InjectRepository(EULA)
    private readonly eulaRepository: Repository<EULA>,
    private readonly jwtService: JwtService,
    @InjectPinoLogger(AuthService.name) private readonly logger: PinoLogger
  ) {}

  async checkIfFirstLogin() {
    try {
      this.logger.info('Checking if first login');

      const user = await this.userRepository.findOneBy({ username: 'admin' });

      this.logger.info({ user }, 'User');
      if (!user) {
        return false;
      }

      const bcrypt = new Bcrypt();
      return await bcrypt.compare('admin', user.hashedPassword);
    } catch (err) {
      this.logger.error({ err }, 'Error while checking if first login');

      throw new Error('Failed to check if first login');
    }
  }

  async updateUserPassword(
    userId: number,
    newPassword: string,
    currentUserId: number
  ): Promise<User> {
    try {
      this.logger.info(
        { userId, newPassword, currentUserId },
        'Updating user password'
      );

      let user = await this.userRepository.findOneBy({ id: userId });

      this.logger.info({ user }, 'User');
      if (!user) {
        throw new NotFoundException('User not found');
      }

      const bcrypt = new Bcrypt();
      const hashedPassword = await bcrypt.hash(newPassword);

      user = await this.userRepository.save({
        ...user,
        hashedPassword,
        updatedByUid: currentUserId,
      });

      this.logger.info({ user }, 'Updated user');

      return user;
    } catch (err) {
      this.logger.error(
        { err, userId, newPassword, currentUserId },
        'Error while updating user password'
      );

      if (err?.response?.statusCode) {
        throw err;
      }

      throw new Error('Failed to update user password');
    }
  }

  async resetAdminPassword(resetPasswordDto: ResetPasswordDto): Promise<User> {
    try {
      this.logger.info({ resetPasswordDto }, 'Resetting admin password');

      const firstLogin = await this.checkIfFirstLogin();

      this.logger.info({ firstLogin }, 'First login');
      if (!firstLogin) {
        throw new BadRequestException('Admin password already updated');
      }

      let user = await this.userRepository.findOneBy({ username: 'admin' });

      this.logger.info({ user }, 'User');
      if (!user) {
        throw new NotFoundException('User not found');
      }

      user = await this.updateUserPassword(
        user.id,
        resetPasswordDto.newPassword,
        new User().id
      );

      this.logger.info({ user }, 'Updated user');

      return user;
    } catch (err) {
      this.logger.error(
        { err, resetPasswordDto },
        'Error while resetting admin password'
      );

      if (err?.response?.statusCode) {
        throw err;
      }

      throw new Error('Failed to reset admin password');
    }
  }

  async authenticateUser(loginDto: LoginDto) {
    try {
      this.logger.info({ loginDto }, 'Authenticating user');

      const user = await this.userRepository.findOne({
        where: {
          username: loginDto.username,
        },
      });

      this.logger.info({ user }, 'User');
      if (!user) {
        throw new NotFoundException('User not found');
      }

      const bcrypt = new Bcrypt();
      const isPasswordValid = await bcrypt.compare(
        loginDto.password,
        user.hashedPassword
      );

      this.logger.info({ isPasswordValid }, 'Is Password valid');
      if (!isPasswordValid) {
        throw new UnauthorizedException('Incorrect username or password');
      }

      if (!user.active) {
        throw new BadRequestException(
          'Cant login in active user, contact admin'
        );
      }

      const eula = await this.eulaRepository.findOne({ where: { id: 1 } });

      if (!eula.accepted) {
        throw new ForbiddenException('EULA is not accepted');
      }

      const payload = {
        user_id: user.id,
        username: user.username,
        role: user.role,
      };

      this.logger.info({ payload }, 'Payload');

      return {
        access_token: await this.jwtService.signAsync(payload),
      };
    } catch (err) {
      this.logger.error({ err, loginDto }, 'Error while authenticating user');

      if (err?.response?.statusCode) {
        throw err;
      }

      throw new Error('Failed to authenticate user');
    }
  }

  async resetPassword(
    userId: number,
    resetPasswordDto: ResetPasswordDto,
    currentUserId: number
  ): Promise<User> {
    try {
      this.logger.info({ userId, resetPasswordDto }, 'Resetting user password');

      let user = await this.userRepository.findOneBy({ id: userId });

      this.logger.info({ user }, 'User');
      if (!user) {
        throw new NotFoundException('User not found');
      }

      user = await this.updateUserPassword(
        user.id,
        resetPasswordDto.newPassword,
        currentUserId
      );

      this.logger.info({ user }, 'Updated user');

      return user;
    } catch (err) {
      this.logger.error(
        { err, userId, resetPasswordDto },
        'Error while resetting user password'
      );

      if (err?.response?.statusCode) {
        throw err;
      }

      throw new Error('Failed to reset user password');
    }
  }
}

File: src/backend/app/modules/auth/auth_service.py
from sqlalchemy.ext.asyncio import AsyncSession
from app.modules.user.user_service import get_user_by_username, update_user_password, get_user_by_id
from app.core.security import verify_password, create_access_token, create_refresh_token, decode_token
from app.core.logger import logger
from fastapi import HTTPException
from app.modules.user.models.user import User
from app.modules.licenses.licesses_service import validate_license


async def authenticate_user(db: AsyncSession, username: str, password: str):
    user = await get_user_by_username(db, username)
    if not user or not verify_password(password, user.hashed_password):
        logger.warning(f"Authentication failed for user: {username}")
        return False

    if not user.active:
        raise HTTPException(
            status_code=400,
            detail="Cant login in active user, contact admin")

    return user


async def create_tokens_for_user(user):
    access_token = create_access_token(
        {"username": user.username, "role": user.role, "user_id": user.id})
    return access_token
    # refresh_token = create_refresh_token({"username": user.username, "role": user.role})
    # return access_token, refresh_token


async def verify_and_refresh_token(db: AsyncSession, refresh_token: str):
    payload = decode_token(refresh_token)
    if not payload:
        return None

    username = payload.get("username")
    if not username:
        return None

    user = await get_user_by_username(db, username)
    if not user:
        return None

    access_token = create_access_token(
        {"username": user.username, "role": user.role})
    new_refresh_token = create_refresh_token(
        {"username": user.username, "role": user.role})
    return {"access_token": access_token, "refresh_token": new_refresh_token}

# First Login APIs


async def check_if_first_login(db: AsyncSession):
    user = await get_user_by_username(db, "admin")
    if not user:
        return False

    if verify_password("admin", user.hashed_password):
        return True
    return False


async def reset_admin_password(db: AsyncSession, new_password: str):
    first_login = await check_if_first_login(db)
    if not first_login:
        raise HTTPException(status_code=400,
                            detail="Admin password already updated")

    user = await get_user_by_username(db, "admin")
    if not user:
        raise HTTPException(status_code=404, detail="User not found")

    user = await update_user_password(db, user.id, new_password, User())
    logger.info(f"Password reset for user: admin")
    if not user:
        return False
    return True


async def reset_password(
        db: AsyncSession,
        token: str,
        new_password: str,
        current_user: User):
    payload = decode_token(token)
    if not payload:
        raise HTTPException(status_code=400, detail="Invalid or expired token")

    username = payload.get("username")
    if not username:
        raise HTTPException(status_code=400, detail="Invalid token")

    user = await get_user_by_username(db, username)
    if not user:
        raise HTTPException(status_code=404, detail="User not found")

    user = await update_user_password(db, user.id, new_password, current_user)
    logger.info(f"Password reset for user: {username}")
    if not user:
        return False
    return True


async def reset_password_user_id(
        db: AsyncSession,
        user_id: int,
        new_password: str,
        current_user: User):
    user = await get_user_by_id(db, user_id)
    if not user:
        raise HTTPException(status_code=404, detail="User not found")

    user = await update_user_password(db, user.id, new_password, current_user)
    logger.info(f"Password reset for user: {user_id}")
    if not user:
        return False
    return True
